src/root/NetworkAgnostic.ts

```diff
@@ -0,0 +1,116 @@
+  { name: 'verifyingContract', type: 'address' },
+]
+
+const metaTransactionType = [
+  { name: 'nonce', type: 'uint256' },
+  { name: 'from', type: 'address' },
+  { name: 'functionSignature', type: 'bytes' },
+]
+
+export default class NetworkAgnostic extends ContractsBase {
+  web3Client: Web3Client
+
+  constructor(web3Client: Web3Client) {
+    super(web3Client)
+    this.web3Client = web3Client
+  }
+
+  async getTokenContract(tokenAddress: address) {
+    const contract = new this.web3Client.networkAgnosticWeb3.eth.Contract(ChildTokenArtifact.abi, tokenAddress)
+    return contract
+  }
+
+  async getContract(contractABI: any, contractAddress: address) {
+    const contract = new this.web3Client.networkAgnosticWeb3.eth.Contract(contractABI, contractAddress)
+    return contract
+  }
+  async transfer(tokenAddress: address, recipientAddress: address, amount: BN, user: address | string) {
+    const contract = await this.getTokenContract(tokenAddress)
+    let functionSignature = await contract.methods.transfer(recipientAddress, this.encode(amount)).encodeABI()
+    await this.networkAgnosticTransaction(tokenAddress, contract, functionSignature, user)
+  }
+
+  async approve(tokenAddress: address, spender: address, amount: BN, user: address) {
+    const contract = await this.getTokenContract(tokenAddress)
+    let functionSignature = await contract.methods.approve(spender, this.encode(amount)).encodeABI()
+    await this.networkAgnosticTransaction(tokenAddress, contract, functionSignature, user)
+  }
+
+  async networkAgnosticTransaction(
+    contractAddress: address,
+    contract: Contract,
+    functionSignature: any | string,
+    userAddress: address
+  ) {
+    const tokenName = await contract.methods.name().call()
+    const parentChainId = '3' // chain id of the network tx is signed on
+    let domainData = {
+      name: tokenName,
+      version: '1',
+      chainId: parentChainId,
+      verifyingContract: contractAddress,
+    }
+    let nonce = await contract.methods.getNonce(userAddress).call()
+
+    let message = {
+      nonce: parseInt(nonce),
+      from: userAddress,
+      functionSignature: functionSignature,
+      network: 'Interact with Matic Network',
+    }
+    const dataToSign = JSON.stringify({
+      types: {
+        EIP712Domain: domainType,
+        MetaTransaction: metaTransactionType,
+      },
+      domain: domainData,
+      primaryType: 'MetaTransaction',
+      message: message,
+    })
+    return await this.web3Client.parentWeb3.eth.currentProvider.send(
+      {
+        jsonrpc: '2.0',
+        id: 999999999999,
+        method: 'eth_signTypedData_v4',
+        params: [userAddress, dataToSign],
+      },
+      async (...response) => {
+        let { r, s, v } = await this.getSignatureParameters(response[1].result)
+        await contract.methods.executeMetaTransaction(userAddress, functionSignature, r, s, v).send({
+          from: userAddress,
+        })
+      }
+    )
+  }
```

Comment:
You will have to return a promise that resolves in the callback. Current implementation will not wait for callback before returning flow to caller.